src/mozanalysis/experiment.py

```diff
@@ -0,0 +1,508 @@
+            factor '7' removes weekly seasonality, and the `+1` accounts
+            for the fact that enrollment typically starts a few hours
+            before UTC midnight.
+        addon_version (str): The version of the experiment addon. Some
+            addon experiment slugs get reused - in those cases we need to
+            filter on the addon version also.
+    """
+    def __init__(
+        self, experiment_slug, start_date, num_dates_enrollment=None,
+        addon_version=None
+    ):
+        # Let's be conservative about storing state - it doesn't belong
+        # in this class. Treat these attributes as immutable.
+        # These attributes are stored because it would be a PITA not to
+        # store them:
+        #   - they are required by both `get_enrollments()` and
+        #       `get_per_client_data()`
+        #   - if you were accidentally inconsistent with their values
+        #       then you would hit trouble quickly!
+        self.experiment_slug = experiment_slug
+        self.start_date = start_date
+        self.num_dates_enrollment = num_dates_enrollment
+        self.addon_version = addon_version
+
+    def get_enrollments(self, spark, study_type='pref_flip', end_date=None):
+        """Return a DataFrame of enrolled clients.
+
+        This works for pref-flip and addon studies.
+
+        The underlying queries are different for pref-flip vs addon
+        studies, because as of 2019/04/02, branch information isn't
+        reliably available in the `events` table for addon experiments:
+        branch may be NULL for all enrollments. The enrollment
+        information for them is most reliably available in
+        `telemetry_shield_study_parquet`. Once this issue is resolved,
+        we will probably start using normandy events for all desktop
+        studies.
+        Ref: https://bugzilla.mozilla.org/show_bug.cgi?id=1536644
+
+        Args:
+            spark: The spark context.
+            study_type (str): One of the following strings:
+                - 'pref_flip'
+                - 'addon'
+            end_date (str, optional): Ignore enrollments after this
+                date: for faster queries on stale experiments. If you
+                set `num_dates_enrollment` then do not set this; at best
+                it would be redundant, at worst it's contradictory.
+
+        Returns:
+            A Spark DataFrame of enrollment data. One row per
+            enrollment. Columns:
+                - client_id (str)
+                - enrollment_date (str): e.g. '20190329'
+                - branch (str)
+        """
+        if study_type == 'pref_flip':
+            enrollments = self._get_enrollments_view_normandy(spark)
+
+        elif study_type == 'addon':
+            enrollments = self._get_enrollments_view_addon(spark, self.addon_version)
+
+        # elif study_type == 'glean':
+        #     raise NotImplementedError
```

Comment:
Is the `df` argument to `get_per_client_data` the other thing we think we'd need to change to make something Gleanlike work?

Reply:
IIRC Glean is going to send enrollment events promptly - I think to the `events` table. In the long term it will probably be worth making a `glean_enrollments` table in the ETL but in the short term we'd probably just filter the events table? So we'd add a method `_get_enrollments_view_glean()`, and call it from this point in `get_enrollments()`.

The implementation of `get_per_client_data` shouldn't have to change - the user would just supply Glean's equivalent to `main_summary` as `df` (which I think is the part that directly answers your question?)